coreos-koji-tagger/coreos_koji_tagger.py

```diff
@@ -393,6 +394,41 @@ def process_lockfiles(self, commit=None):
                             poll_interval=10)
                 logger.info('Tagging done')
 
+                # Subsequently run a distrepo task because there are
+                # races in tag2distrepo. https://pagure.io/koji/issue/1630
+                # Before running distrepo let's wait for all rpms to
+                # pass through signing and make it into the target tag
+                while True:
+                    currentbuildids = self.get_tagged_buildids(self.target_tag)
+                    difference = desiredbuildids - currentbuildids
+                    if difference:
+                        logger.info('Waiting on builds to be signed')
+                        logger.info('Remaining builds: %s' %
+                                        [buildsinfo[x].nvr for x in difference])
+                        time.sleep(5)
+                        continue
+                    break
+                # This code is mostly stolen from:
+                # https://pagure.io/releng/tag2distrepo/blob/master/f/tag2distrepo.py
+                taginfo = self.koji_client.getTag(self.target_tag)
+                keys = taginfo['extra'].get("tag2distrepo.keys", '').split()
+                task_opts = {
+                    'arch': taginfo['arches'].split(),
+                    'comp': None,
+                    'delta': [],
+                    'event': None,
+                    'inherit': False,
+                    'latest': False,
+                    'multilib': False,
+                    'split_debuginfo': False,
+                    'skip_missing_signatures': False,
+                    'allow_missing_signatures': False,
+                }
+                task = self.koji_client.distRepo(self.target_tag,
+                                                    keys, **task_opts)
+                watch_tasks(self.koji_client, [task], poll_interval=10)
```

Comment:
One improvement here is to split out waiting for the tags and doing `distRepo` on a separate thread so that the processing time for each (relevant) message stays low. Otherwise, there's a theoretical issue with messages queuing up faster than we're clearing them. I don't think we'll hit this in practice today since we don't have super high churn on the lockfiles.

Reply:
I think right now we want to keep it so that requests are processed serially so that more than one distrepo can't be running at the same time. Do you agree?

Keeping message processing time low would be nice but I think your proposal could introduce a race condition again, right?

Reply:
Right yeah, it'd require careful implementation so we don't have a race. What I was thinking was a shared list object where the tagger just appends the rpms to wait for, and the thread would just take out of the list, wait for them to enter the pool, then run `distRepo`. (And e.g. the tagger could add more rpms on the list while the waiting or `distRepo` is happening for the next iteration on the thread main loop).

Reply:
To be clear, I think it'd be cool, but I definitely don't think it's worth blocking this!

Reply:
Yeah that seems interesting, but probably better to keep this simple especially since we don't have to worry about lost messages because we're sitting on top of AMQP already. We can let AMQP handle the complexity. The only real benefit I see to adding complexity here would be to improve the performance of the koji tagger. Right now our bodhi run (the input to the lockfiles getting changed) takes a long time (much longer than the koji tagger) so I'd say it's a problem we won't really see until we start updating more and more lockfiles.